### PR TITLE
Update Spec for Explainers

### PR DESCRIPTION
**Pull request by cliveseldon**
Provides an update to the spec to allow explainers to be specified.
As this creates a new level in the spec will need changes to existing code/yaml once agreed.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/237)

### CODE REVIEW

**Comment by cliveseldon** [comment]
Can I get your review on this explainer spec as broadly discussed in the weekly meeting yesterday.
@ellis-bigelow @rakelkar @animeshsingh @yuzisun 

As we agreed Explainers can be canaried and also should not interact with min/max replicas specified for predictors a new level in the spec hierarchy has been created.

**Comment by cliveseldon** [comment]
From conversations with @ellis-bigelow  I've updated to make the spec backwards compatible for now. So we can move to breaking changes in v1alpha2.

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 1]
```diff
@@ -42,6 +41,29 @@ type ModelSpec struct {
...
+	// The kind of Alibi explanation server to create, e.g. AnchorTabular
+	Type string `json:"type"`
+	// The location of a fit explanation model
+	SavedExplainerURI string `json:"savedExplainerUri,omitempty"`
+	// Defaults to latest Alibi Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	// Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// Inline custom parameter settings for explainer
+	Config v1.ConfigMap `json:"config,omitempty"`
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: I'm pretty sure configmap is the wrong concept here. Can you provide another example of someone passing a configmap through an interface like this? I think the better abstraction is probably a map<string,string>. 

Can you provide an example of what types of parameters would be passed through here?

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 2]
```diff
@@ -42,6 +41,29 @@ type ModelSpec struct {
 	XGBoost    *XGBoostSpec    `json:"xgboost,omitempty"`
 	SKLearn    *SKLearnSpec    `json:"sklearn,omitempty"`
 	PyTorch    *PyTorchSpec    `json:"pytorch,omitempty"`
+	Explain    *ExplainerSpec  `json:"explain,omitempty"`
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: I think this is at the right level. We will probably want to push the 1-of predictors down a level under `predict`

Comment (animeshsingh) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: sorry for missing yesterday's meeting. Conceptually, I am having hard time convincing why would Explainer Spec be at the same level as specs for 'Model Types' e.g. PyTorch, or Custom Model?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (animeshsingh) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: What inputs we are expecting in Map? Also this sentence _'AlibiExplainerSpec defines the arguments for configuring an Alibi explanation serving'_ - not sure if its coming across in a cohesive way?

With this _'Alibi explanation server'_ -  I believe then the agreement yesterday was that explainer is another server in KFServing spec? Tomorrow if we have 'Bias Detector', I presume that will be another server on same lines?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 2 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: @animeshsingh The core proposals are in this [doc](https://docs.google.com/document/d/1V-ATKWEuV4-uqkc6CqMsgmkTFt-2Sf1wBofpBE1bqBI/edit) discussed at the meeting. The main one for level is that Explainers can load saved models (some need to be fit on the training data) and they can be canaried (as they may fail). Eventually we would put the Predictor specs as @ellis-bigelow mentioned in a level further down to separate them more clearly - see my earlier commit.

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: @ellis-bigelow Some example parameters for AnchorTabular can be seen [here](https://docs.seldon.io/projects/alibi/en/v0.2.2/api/alibi.explainers.anchor_tabular.html#alibi.explainers.anchor_tabular.AnchorTabular.explain). 

A Map is also probably fine.

Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: @animeshsingh We are handling explainers in a synchronous way only so any other component that can be used in this way could also be added eventually (see comment on levels below).

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 2 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: @ellis-bigelow Are you suggesting to push the other predictors down a level now? I thought the idea was to not break the YAML specs at present?

Comment (hougangliu) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: s/Explain/Explainer
or
s/ExplainerSpec/ExplainSpec

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 3]
```diff
@@ -42,6 +41,29 @@ type ModelSpec struct {
...
+	Custom *CustomSpec         `json:"custom,omitempty"`
+}
+
+// AlibiExplainerSpec defines the arguments for configuring an Alibi explanation serving
+type AlibiExplainerSpec struct {
+	// The following fields follow a "1-of" semantic. Users must specify exactly one of Type or ExplainerURI.
+	// The kind of Alibi explanation server to create, e.g. AnchorTabular
+	Type string `json:"type"`
+	// The location of a fit explanation model
+	SavedExplainerURI string `json:"savedExplainerUri,omitempty"`
```
Comment (hougangliu) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: is SavedExplainerURI used to point an existing explainer service url?

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (hougangliu) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I think you can replace Config v1.ConfigMap as `Parameters []ParameterSpec`

type ParameterSpec struct {
  key String
  value String
}

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 3 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: No, to a saved explainer model. So simialr to predictors that can load their modelWeights from a bucket so can an explainer.

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 2 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I think `explain` is matching the future `predict` (see first comment). It could indeed also be `explainer` and `predictor`.

Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: Updated

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 3 above)
Comment (hougangliu) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I see. Thanks.

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 2 above)
Comment (yuzisun) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I think it is worth considering using v1alpha2 if we do not want to break existing yaml, patching to existing v1alpha1 might be confusing to user. We can do conversion from v1alpha1 to v1alpha2 in mutation webhook if we really want to back support v1alpha1, I believe knative does this too between v1alpha1 and v1beta1 .

Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I have added a comment to the spec to make it clear the current location of `explain`. I think the advantage of keeping it at v1alpha1 is that we can get a working example and leave the v1alpha2 changes to a separate piece of work. Doing them together will be a bigger PR. 

I'm happy either way. Just want to keep momentum.

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (ellis-bigelow) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: @cliveseldon, should we just wire all those parameters from  https://docs.seldon.io/projects/alibi/en/v0.2.2/api/alibi.explainers.anchor_tabular.html#alibi.explainers.anchor_tabular.AnchorTabular.explain? We could even allow a map for kwargs.

**Comment by ellis-bigelow** [comment]
This looks good to me Clive!

Would love consensus from the group and then we can merge.

File: pkg/apis/serving/v1alpha1/kfservice_types.go (hunk 1 above)
Comment (cliveseldon) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: I think for now we can stick with a map. Once things settle we could consider maybe the most important explicitly. The issue is there could be a large number which may also change more frequently that the rest of the Spec here.

**Comment by rakelkar** [comment]
Looks good to me - is it fair to assume that when the explainer is instantiated the model URI will be passed to it? Will be interesting to see this evolve with v1alpha2 :)

**Comment by cliveseldon** [comment]
@rakelkar  Yes the predict endpoint will be passed so "black box" explainers can access the internal service point to make calls.

**Comment by ellis-bigelow** [comment]
I wonder if we should decide a set of environment variables passed to all
components like:

Predict uri
Storage uri
Explain uri

Etc.

That way, any custom spec will have the power to utilize the graph however
it needs.

**Comment by rakelkar** [comment]
Sounds reasonable to me...

**Comment by animeshsingh** [comment]
> @rakelkar Yes the predict endpoint will be passed so "black box" explainers can access the internal service point to make calls.

How do we define "black box" eplainers here?

**Comment by animeshsingh** [comment]
@ellis-bigelow @cliveseldon 

its looking good to me, though any chance we can get a diagrammatic representation of the spec tree? where we settle for now, and what will change with alpha2? Will make these conversatoins much smoother

**Comment by rakelkar** [comment]
Will also be good to get an example of whitebox.. that might help
inform/influence #222

**Comment by cliveseldon** [comment]
@animeshsingh 

Probably easiest to see in an example. For v1alpha1 for an sklearn model with explainer then non-breaking change it will be like:
```
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "sklearn-iris"
spec:
  default:
    sklearn:
      modelUri: "gs://kfserving-samples/models/sklearn/iris"
    explain:
      alibi:
        type: "anchor_tabular"
        savedExplainerUri: "gs://kfserving-samples/models/sklearn/iris_explainer"
```

The v1alpha2 proposal is to change to:

```
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "sklearn-iris"
spec:
  default:
    predict:
      sklearn:
        modelUri: "gs://kfserving-samples/models/sklearn/iris"
    explain:
      alibi:
        type: "anchor_tabular"
        savedExplainerUri: "gs://kfserving-samples/models/sklearn/iris_explainer"
```

**Comment by cliveseldon** [comment]
@animeshsingh whitebox should be the same. Blackbox explainers need to be autowired with the predict URI while whitebox need the model URI. Both are available to kfserving at reconcile time.

**Comment by rakelkar** [comment]
Can you draft an example of what that might look like? How would you expect
the model initializer to kick in?

**Comment by cliveseldon** [comment]
@rakelkar 
The modelInitializer would work identically to how it works for predictors. The Alibi explainer takes a command line arg for the savedexplainerUri location in the same way as our built in predictors do for modelUri. So the same mutation logic can be run.

**Comment by ellis-bigelow** [comment]
@rakelkar, if we detect a storageURI for any spec (explain/predict/preprocess), we should mount up that DIR and get it local.

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 4]
```diff
@@ -42,6 +41,31 @@ type ModelSpec struct {
...
+	Custom *CustomSpec       `json:"custom,omitempty"`
+}
+
+// AlibiExplainSpec defines the arguments for configuring an Alibi Explanation Server
+type AlibiExplainSpec struct {
+	// The following fields follow a "1-of" semantic. Users must specify exactly one of Type or ExplainerURI.
+	// The kind of Alibi explanation server to create, e.g. AnchorTabular
+	Type string `json:"type"`
+	// The location of a fit explanation model
+	SavedExplainerURI string `json:"savedExplainerUri,omitempty"`
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: StorageURI

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 5]
```diff
@@ -42,6 +41,31 @@ type ModelSpec struct {
...
+	// Defaults to latest Alibi Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	// Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// Inline custom parameter settings for explainer
+	Config map[string]string `json:"config,omitempty"`
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: nice!

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 6]
```diff
@@ -42,6 +41,31 @@ type ModelSpec struct {
...
+	// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+	Alibi  *AlibiExplainSpec `json:"alibi,omitempty"`
+	Custom *CustomSpec       `json:"custom,omitempty"`
+}
+
+// AlibiExplainSpec defines the arguments for configuring an Alibi Explanation Server
+type AlibiExplainSpec struct {
+	// The following fields follow a "1-of" semantic. Users must specify exactly one of Type or ExplainerURI.
+	// The kind of Alibi explanation server to create, e.g. AnchorTabular
+	Type string `json:"type"`
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: Can you make an enum for this? We can always add values in a backwards compat way.

**Comment by animeshsingh** [comment]
Can we get at-least one sample - specifically for a feature like this, that gives the path for testing?

**Comment by cliveseldon** [comment]
There can only be an example once the controller-manager is updated which will need to be done after this is merged. We could create a example based on the existing example in https://github.com/kubeflow/kfserving/tree/master/docs/samples/explanation/income but the yaml would be unusable until the controller-manager work is done.

**Comment by cliveseldon** [comment]
Think this is ready to merge.
